kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,426 @@
+///    so by transforming the actions using [`get_add_transform_expr`], and generating selection
+///    vectors with the following rules:
+///     - If a `cdc` action was found in the prepare phase, only `cdc` actions are selected
+///     - Otherwise, select `add` and `remove` actions. Note that only `remove` actions that do not
+///       share a path with an `add` action are selected.
+///
+/// Note: As a consequence of the two phases, LogReplayScanner will iterate over each action in the
+/// commit twice. It also may use an unbounded amount of memory, proportional to the number of
+/// `add` + `remove` actions in the _single_ commit.
+struct LogReplayScanner {
+    // True if a `cdc` action was found after running [`LogReplayScanner::prepare_table_changes`]
+    has_cdc_action: bool,
+    // A map from path to the deletion vector from the remove action. It is guaranteed that there
+    // is an add action with the same path in this commit
+    remove_dvs: HashMap<String, DvInfo>,
+    // The commit file that this replay scanner will operate on.
+    commit_file: ParsedLogPath,
+    // The timestamp associated with this commit. By default this is the file modification time
+    // from the commit's [`FileMeta`]. If there is a [`CommitInfo`] with a timestamp generated by
+    // in-commit timestamps, that timestamp will be used instead.
+    //
+    // Note: This will be used once an expression is introduced to transform the engine data in
+    // [`TableChangesScanData`]
+    #[allow(unused)]
+    timestamp: i64,
+    // The data skipping filter for filtering log actions
+    filter: Option<DataSkippingFilter>,
+}
+
+fn prepare_table_changes(
+    commit_file: ParsedLogPath,
+    engine: &dyn Engine,
+    table_schema: &SchemaRef,
+    physical_predicate: Option<ExpressionRef>,
+) -> DeltaResult<LogReplayScanner> {
+    let filter = DataSkippingFilter::new(engine, table_schema, physical_predicate);
+    let visitor_schema = PreparePhaseVisitor::schema()?;
+    let action_iter = engine.get_json_handler().read_json_files(
+        &[commit_file.location.clone()],
+        visitor_schema,
+        None,
+    )?;
+
+    let mut remove_dvs = HashMap::default();
+    let mut add_paths = HashSet::new();
+    let mut has_cdc_action = false;
+    let mut timestamp = commit_file.location.last_modified;
+    for actions in action_iter {
+        let actions = actions?;
+        let selection_vector = match &filter {
+            Some(filter) => filter.apply(actions.as_ref())?,
+            None => vec![true; actions.len()],
+        };
+
+        let mut visitor = PreparePhaseVisitor::new(
+            selection_vector,
+            &mut add_paths,
+            &mut remove_dvs,
+            &mut has_cdc_action,
+            &mut timestamp,
+        );
+        visitor.visit_rows_of(actions.as_ref())?;
+
+        if let Some(protocol) = visitor.protocol {
+            protocol.ensure_read_supported()?;
+        }
```

Comment:
this is saying that _if_ a protocol action is present during iteration we check it's supported (I read the comment above - just trying to think through this)

do we need to check for table features at all? I guess CDF doesn't need to? (and i see the table properties check below)

Reply:
> this is saying that if a protocol action is present during iteration we check it's supported

Correct

> do we need to check for table features at all? 

No specific reader or writer features.

Reply:
> CDF is absolutely impacted by table features -- DV and ICT at a minimum

Is there any specific behaviour that I should be checking for this? I don't see anything explicit in the delta-spark code for this

Reply:
It's already implemented... just without checking for the table feature:
* ICT - get timestamp from commit info
* DV - use deletion vectors

Reply:
ICT table property checking is currently unsupported, and @zachschuermann  is [working on introducing it ](https://github.com/delta-io/delta-kernel-rs/pull/558)

We actually always have to handle deletion vectors, even when it's not specified in TableProperties. From [the protocol](https://github.com/delta-io/delta-kernel-rs/pull/558): 
> Readers must read the table considering the existence of DVs, even when the delta.enableDeletionVectors table property is not set.